### ISSUE TITLE
Globals not set by default for eval() when embedding

### ISSUE THREAD

**Issue by gelldur** [pybind/pybind11#887]
Make sure you've completed the following steps before submitting your issue -- thank you!
You can remove this template text afterward.

1. Check if your question has already been answered in the [FAQ](http://pybind11.readthedocs.io/en/latest/faq.html) section.
2. Make sure you've read the [documentation](http://pybind11.readthedocs.io/en/latest/). Your issue may be addressed there.
3. If those resources didn't help and you only have a short question (not a bug report), consider asking in the [Gitter chat room](https://gitter.im/pybind/Lobby).
4. If you have a genuine bug report or a more complex question which is not answered in the previous items (or not suitable for chat), please fill in the details below.
5. Include a self-contained and minimal piece of code that reproduces the problem. If that's not possible, try to make the description as clear as possible.

#### Issue description

Example from documentation isn't working
http://pybind11.readthedocs.io/en/master/advanced/embedding.html#getting-started

```
    py::exec(R"(
        kwargs = dict(name="World", number=42)
        message = "Hello, {name}! The answer is {number}".format(**kwargs)
        print(message)
    )");
```
Error:
```
terminate called after throwing an instance of 'pybind11::error_already_set'
  what():  NameError: name 'dict' is not defined

At:
  <string>(3): <module>
```

This also isn't working:
```
	py::exec(R"(
        #kwargs = {'name':"World", 'number':42}
        #message = "Hello, {name}! The answer is {number}".format(**kwargs)
        print("Hello")
    )");
```
Error:
```
terminate called after throwing an instance of 'pybind11::error_already_set'
  what():  NameError: name 'print' is not defined

At:
  <string>(5): <module>
```

#### Reproducible example code

```cpp
#include <iostream>

#include <pybind11/embed.h> // everything needed for embedding

namespace py = pybind11;

int main()
{
	py::scoped_interpreter guard{}; // start the interpreter and keep it alive

	py::exec(R"(
        kwargs = dict(name="World", number=42)
        message = "Hello, {name}! The answer is {number}".format(**kwargs)
        print(message)
    )");

	py::exec(R"(
        #kwargs = {'name':"World", 'number':42}
        #message = "Hello, {name}! The answer is {number}".format(**kwargs)
        print("Hello")
    )");
	return 0;
}
```

**Comment by jagerman** [comment]
Those `dict`s should be `py::dict`s.

**Comment by gelldur** [comment]
@jagerman so bug in documentation yeah ? but what about `print` ?

**Comment by jagerman** [comment]
Yeah, `py::print` as well.  I'll fix the example.

**Comment by jagerman** [comment]
Er, never mind that's not a c++ call.

**Comment by jagerman** [comment]
The print is probably a matter of using Python 2 rather than 3.

**Comment by gelldur** [comment]
Even if i set python 3.6 in cmake still doesn't work :(

Checked even for python2
```
	py::exec(R"(
        #kwargs = {'name':"World", 'number':42}
        #message = "Hello, {name}! The answer is {number}".format(**kwargs)
        print "Hello"
    )");
```

```terminate called after throwing an instance of 'pybind11::error_already_set'
  what():  SyntaxError: ('invalid syntax', ('<string>', 5, 13, 'print "Hello"\n'))```

**Comment by jagerman** [comment]
Confirmed.  Cc @dean0x7d: it seems as though when called for an embedded interpreter `PyEval_GetGlobals` returns `NULL`, because there is no current execution frame.

Adding `py::globals` to the example will fix it, but that feels to me like something that should be the default: perhaps in `eval()`, if `globals` is unset and `PyEval_GetGlobals()` returns `NULL`, we should fall back to `py::globals()` instead of an empty `dict()`?

**Comment by dean0x7d** [comment]
Resolved in #889. Thanks for reporting the issue @gelldur!
